libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/report/BuildMetricsReporterService.kt

```diff
@@ -38,70 +38,88 @@ abstract class BuildMetricsReporterService : BuildService<BuildMetricsReporterSe
 
     private val log = Logging.getLogger(this.javaClass)
 
-    //Task path to build metrics
-    private val buildMetricsMap = ConcurrentHashMap<String, BuildMetricsReporter>()
-    private val taskRecords = ConcurrentHashMap<String, TaskRecord>()
-    private val failureMessages = ConcurrentLinkedQueue<String?>()
-    private val taskClass = ConcurrentHashMap<String, String>()
+    // Tasks and transforms' records
+    private val buildOperationRecords = ConcurrentLinkedQueue<BuildOperationRecord>()
+    private val failureMessages = ConcurrentLinkedQueue<String>()
 
-    open fun add(taskPath: String, type: String, metrics: BuildMetricsReporter) {
-        if (buildMetricsMap.containsKey(taskPath)) {
-            log.warn("Duplicate path $taskPath")
+    // Info for tasks only
+    private val taskPathToMetricsReporter = ConcurrentHashMap<String, BuildMetricsReporter>()
+    private val taskPathToTaskClass = ConcurrentHashMap<String, String>()
+
+    open fun addTask(taskPath: String, taskClass: Class<*>, metricsReporter: BuildMetricsReporter) {
+        taskPathToMetricsReporter.put(taskPath, metricsReporter).also {
+            check(it == null) { "Duplicate task path: $taskPath" }
+        }
+        taskPathToTaskClass.put(taskPath, taskClass.name).also {
+            check(it == null) { "Duplicate task path: $taskPath" }
         }
-        buildMetricsMap[taskPath] = metrics
-        taskClass[taskPath] = type
+    }
+
+    open fun addTransformMetrics(
+        transformPath: String, transformClass: Class<*>, isKotlinTransform: Boolean, startTimeMs: Long, totalTimeMs: Long,
+        buildMetrics: BuildMetrics, failureMessage: String?
```

Comment:
nit: Formatting looks a bit off.

Reply:
In the IDE, it looks like this:
```
    open fun addTransformMetrics(
        transformPath: String, transformClass: Class<*>, isKotlinTransform: Boolean, startTimeMs: Long, totalTimeMs: Long,_____18 blank chars from margin____
        buildMetrics: BuildMetrics, failureMessage: String?
    ) {
```

so it's fine right?

Reply:
Each param should typically be in a separate line if it does not fit in a single line - https://kotlinlang.org/docs/coding-conventions.html#functions

Reply:
The IDE formatter didn't complain, but yeah I updated it.